### ISSUE TITLE
RenderBinding causing concurrent modification during iteration error

### ISSUE THREAD

**Issue by jwehrle** [flutter/flutter#103516]
Getting the following error which appears to have nothing to do with my own source code. At least, there is no reference in the stack to my source code.
```
E/flutter (29940): [ERROR:flutter/lib/ui/ui_dart_state.cc(209)] Unhandled Exception: Concurrent modification during iteration: Instance(length:3) of '_GrowableList'.
E/flutter (29940): #0      ListIterator.moveNext (dart:_internal/iterable.dart:336:7)
E/flutter (29940): #1      SchedulerBinding.handleDrawFrame (package:flutter/src/scheduler/binding.dart:1080:44)
E/flutter (29940): #2      SchedulerBinding._handleDrawFrame (package:flutter/src/scheduler/binding.dart:995:5)
E/flutter (29940): #3      _rootRun (dart:async/zone.dart:1426:13)
E/flutter (29940): #4      _CustomZone.run (dart:async/zone.dart:1328:19)
E/flutter (29940): #5      _CustomZone.runGuarded (dart:async/zone.dart:1236:7)
E/flutter (29940): #6      _invoke (dart:ui/hooks.dart:151:10)
E/flutter (29940): #7      PlatformDispatcher._drawFrame (dart:ui/platform_dispatcher.dart:308:5)
E/flutter (29940): #8      _drawFrame (dart:ui/hooks.dart:115:31)
E/flutter (29940): 

======== Exception caught by scheduler library =====================================================
The following assertion was thrown during a scheduler callback:
'package:flutter/src/rendering/binding.dart': Failed assertion: line 369 pos 12: '!_debugMouseTrackerUpdateScheduled': is not true.


Either the assertion indicates an error in the framework itself, or we should provide substantially more information in this error message to help you determine and fix the underlying cause.
In either case, please report this assertion by filing a bug on GitHub:
  https://github.com/flutter/flutter/issues/new?template=2_bug.md

When the exception was thrown, this was the stack: 
#2      RendererBinding._scheduleMouseTrackerUpdate (package:flutter/src/rendering/binding.dart:369:12)
#3      RendererBinding._handlePersistentFrameCallback (package:flutter/src/rendering/binding.dart:364:5)
#4      SchedulerBinding._invokeFrameCallback (package:flutter/src/scheduler/binding.dart:1144:15)
#5      SchedulerBinding.handleDrawFrame (package:flutter/src/scheduler/binding.dart:1081:9)
#6      SchedulerBinding._handleDrawFrame (package:flutter/src/scheduler/binding.dart:995:5)
#10     _invoke (dart:ui/hooks.dart:151:10)
#11     PlatformDispatcher._drawFrame (dart:ui/platform_dispatcher.dart:308:5)
#12     _drawFrame (dart:ui/hooks.dart:115:31)
(elided 5 frames from class _AssertionError and dart:async)
====================================================================================================
```

Doctor report:

```
[✓] Flutter (Channel stable, 2.10.5, on macOS 12.3.1 21E258 darwin-x64, locale en-US)
    • Flutter version 2.10.5 at /Users/*********/Flutter/flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 5464c5bac7 (3 weeks ago), 2022-04-18 09:55:37 -0700
    • Engine revision 57d3bac3dd
    • Dart version 2.16.2
    • DevTools version 2.9.2

[✓] Android toolchain - develop for Android devices (Android SDK version 29.0.2)
    • Android SDK at /Users/*******/Library/Android/sdk
    • Platform android-31, build-tools 29.0.2
    • Java binary at: /Applications/Android Studio.app/Contents/jre/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 11.0.12+0-b1504.28-7817840)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 13.3.1)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • CocoaPods version 1.11.3

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 2021.2)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 11.0.12+0-b1504.28-7817840)

[✓] VS Code (version 1.40.2)
    • VS Code at *******
    • Flutter extension version 3.8.1

[✓] Connected device (3 available)
    • AOSP on IA Emulator (mobile) • emulator-5554                        • android-x86    • Android 9 (API 28) (emulator)
    • iPhone 13 (mobile)           • *********
com.apple.CoreSimulator.SimRuntime.iOS-15-4 (simulator)
    • Chrome (web)                 • chrome                               • web-javascript • Google Chrome 101.0.4951.64

[✓] HTTP Host Availability
    • All required HTTP hosts are available

• No issues found!
Process finished with exit code 0
```

## Steps to Reproduce

Run a Flutter application. Apparently.

**Comment by darshankawar** [comment]
@jwehrle 
There seems to be 2 different issues from the log you shared.

1. `[ERROR:flutter/lib/ui/ui_dart_state.cc(209)] Unhandled Exception: Concurrent modification during iteration: Instance(length:3) of '_GrowableList'.` 

Indicating that the error is coming from your code implementation itself. See this similar SO link https://stackoverflow.com/questions/22409666/exception-concurrent-modification-during-iteration-instancelength17-of-gr

And response from flutter team member on one such similar issue: https://github.com/flutter/flutter/issues/49339#issuecomment-578662387

Also see https://stackoverflow.com/a/59877613

2. `The following assertion was thrown during a scheduler callback:
'package:flutter/src/rendering/binding.dart': Failed assertion: line 369 pos 12: '!_debugMouseTrackerUpdateScheduled': is not true.`. which is same as discussed in this open issue : https://github.com/flutter/flutter/issues/84241 so I suggest you to follow-up in it for further updates.

Closing this per above. If you disagree or think this is incorrect, write in comments and I'll reopen it.

**Comment by jwehrle** [comment]
@darshankawar thank you for those links as I was at a loss. I have read them over and, looking at my code, I am still confused. Without providing a code dump, let me explain the situation:

I have made a helper flutter package which, among other things uses Overlay.of(context) to insert a list of OverlayEntry. Now, that list is assembled once in a State's initState method, then Overlay.of(context).insertAll() is called once (a bool flag is used to prevent doing this more than once) from SchedulerBinding.instance.addPostFrameCallback. In dispose, all OverlayEntry are removed. Within my flutter package example app this works exactly as expected and the above problem never occurs. There are no warnings. There are no exceptions thrown.

Now, in another of my packages which uses the helper package above, I get the above exception every time the helper package is used.

The list of OverlayEntry is only filled once (a bool flag is used) and only inserted once (same bool flag used). The entries are not removed from the list but each one has its remove method called once in the dispose method. And, as I say, the helper package example app does not have the above problem.

But when and only when the helper package is added does the above exception occur. So there must be some incompatibility that I have not found and I have few clues as to where to look.

**Comment by jwehrle** [comment]
@darshankawar  I have worked around this issue now by avoiding SchedulerBinding. As best I can tell the State of a StatefulWidget was being twice rebuilt in rapid succession and a list member was having addAll called concurrently because of the async nature of SchedulerBinding postFrameCallback. I don't think this should be possible but I am at a loss to see any other explanation. The list in question is declared as a member variable, filled with addAll once in initState, and read in the posFrameCallback.

Whatever the cause, I think the error message could have been far more helpful. It should have at least included a reference to my source code in its stacktrace. And ideally it should have included some more details about the particular exception that was caught.

Also, for anyone else who has similar issues, I got around this by putting all of the OverlayEntries (or what they built, rather) into a Stack.
